fix: validate content-type for refresh token requests

zhyd1997 opened the pull request:
## What does this PR do?

<!-- Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context. List any dependencies that are required for this change. -->
Added a check to ensure the request's content-type is `application/x-www-form-urlencoded`
Returns a 400 error response with an appropriate error message if the validation fails


- Fixes #24964(GitHub issue number)
- Fixes CAL-6707 (Linear issue number - should be visible at the bottom of the GitHub issue description)

## Visual Demo (For contributors especially)

A visual demonstration is strongly recommended, for both the original and new change **(video / image - any one)**.

#### Video Demo (if applicable):

- Show screen recordings of the issue or feature.
- Demonstrate how to reproduce the issue, the behavior before and after the change.

#### Image Demo (if applicable):

- Add side-by-side screenshots of the original and updated change.
- Highlight any significant change(s).

## Mandatory Tasks (DO NOT REMOVE)

- [x] I have self-reviewed the code (A decent size PR without self-review might be rejected).
- [ ] I have updated the developer docs in /docs if this PR makes changes that would require a [documentation change](https://cal.com/docs). If N/A, write N/A here and check the checkbox.
- [ ] I confirm automated tests are in place that prove my fix is effective or that my feature works.

## How should this be tested?

<!-- Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration. Write details that help to start the tests -->

- Are there environment variables that should be set?
- What are the minimal test data to have?
- What is expected (happy path) to have (input and output)?
- Any other important info that could help to test that PR

## Checklist

<!-- Remove bullet points below that don't apply to you -->

- I haven't read the [contributing guide](https://github.com/calcom/cal.com/blob/main/CONTRIBUTING.md)
- My code doesn't follow the style guidelines of this project
- I haven't commented my code, particularly in hard-to-understand areas
- I haven't checked if my changes generate no new warnings

zhyd1997 committed:
fix: validate content-type for refresh token requests

- Added a check to ensure the request's content-type is "application/x-www-form-urlencoded"
- Returns a 400 error response with an appropriate error message if the validation fails

zhyd1997 committed:
Update apps/web/app/api/auth/oauth/refreshToken/route.ts

Co-authored-by: cubic-dev-ai[bot] <191113872+cubic-dev-ai[bot]@users.noreply.github.com>

pallava-joshi reviewed (changes requested):
we have some failing unit tests. can you please look into that.

zhyd1997 committed:
Update apps/web/app/api/auth/oauth/refreshToken/route.ts

Co-authored-by: cubic-dev-ai[bot] <191113872+cubic-dev-ai[bot]@users.noreply.github.com>

pallava-joshi committed:
Merge branch 'main' into fix/issue-24964-1762434985

dhairyashiil committed:
Merge branch 'main' into fix/issue-24964-1762434985

dhairyashiil reviewed (commented):
Hi @zhyd1997, I was going through RFC 6749 and noticed it seems to require both the `authorization_code` and `refresh_token` grant requests to use `application/x-www-form-urlencoded`. If that’s correct, do you think it might make sense to include that here as well? [link](https://github.com/calcom/cal.com/blob/85fe40688e771cb5ea77e46bd30f6da1eb602976/apps/web/app/api/auth/oauth/token/route.ts#L11C1-L11C43)

Also, I saw that the RFC mentions adding `Cache-Control: no-store` and `Pragma: no-cache` headers on all token responses (both success and error). Maybe we could consider adding those too?

zhyd1997 committed:
Merge branch 'main' into fix/issue-24964-1762434985

zhyd1997 committed:
feat(apps/web): add `isValidOAuthTokenHeader` utility and integrate into OAuth token handling

- Introduced a new utility function `isValidOAuthTokenHeader` to validate the content type of OAuth token requests.
- Updated the refresh token and token route handlers to utilize this new validation function, improving request handling and error responses.
- Enhanced error handling for invalid content types in the token route.

zhyd1997 committed:
fix(apps/web): invert validation logic for OAuth token request handling

- Updated the refresh token route to return an error for invalid content types instead of valid ones, improving request validation logic.

zhyd1997 committed:
refactor(apps/web): improve `isValidOAuthTokenHeader` function and update imports

- Enhanced the `isValidOAuthTokenHeader` function to return false for missing content types and improved readability.
- Updated import paths for `isValidOAuthTokenHeader` in the OAuth token and refresh token route handlers for consistency.

zhyd1997 committed:
refactor(apps/web): export VALID_REQ_CONTENT_TYPE and update error messages

- Exported `VALID_REQ_CONTENT_TYPE` from `isValidOAuthTokenHeader` for reuse in OAuth token and refresh token route handlers.
- Updated error messages in the token route handlers to use the exported content type constant, enhancing maintainability and clarity.

pallava-joshi committed:
Merge branch 'main' into fix/issue-24964-1762434985

pallava-joshi committed:
Merge branch 'main' into fix/issue-24964-1762434985
